File: src/lib/sinhala-script.ts
/** Sinhala Unicode block (U+0D80–U+0DFF). */
const SINHALA_CHAR = /[\u0D80-\u0DFF]/g;
const LETTER = /\p{L}/gu;

function countMatches(text: string, pattern: RegExp): number {
  return [...text.matchAll(pattern)].length;
}

/**
 * True when text is meaningfully Sinhala-script:
 * - at least 3 Sinhala block characters, OR
 * - Sinhala letters are >= half of all Unicode letters in the string.
 */
export function hasSinhalaScript(text: string): boolean {
  const trimmed = text.trim();
  if (!trimmed) return false;

  const sinhalaChars = countMatches(trimmed, SINHALA_CHAR);
  if (sinhalaChars >= 3) return true;

  const letters = trimmed.match(LETTER) ?? [];
  if (letters.length === 0) return false;

  const sinhalaLetters = letters.filter((ch) => /[\u0D80-\u0DFF]/.test(ch)).length;
  return sinhalaLetters >= 1 && sinhalaLetters / letters.length >= 0.5;
}

/** Public news listings: title must qualify; excerpt can support short titles. */
export function hasSinhalaNewsText(title: string, excerpt?: string): boolean {
  if (hasSinhalaScript(title)) return true;
  const combined = `${title.trim()} ${(excerpt ?? "").trim()}`.trim();
  return combined !== title.trim() && hasSinhalaScript(combined);
}

File: services/newsbot/sinhala_script.py
"""Detect meaningful Sinhala script in RSS titles (Unicode U+0D80–U+0DFF)."""

from __future__ import annotations

import re

_SINHALA = re.compile(r"[\u0d80-\u0dff]")
_LETTER = re.compile(r"\p{L}", re.UNICODE)


def has_sinhala_script(text: str) -> bool:
    trimmed = (text or "").strip()
    if not trimmed:
        return False

    sinhala_chars = len(_SINHALA.findall(trimmed))
    if sinhala_chars >= 3:
        return True

    letters = _LETTER.findall(trimmed)
    if not letters:
        return False

    sinhala_letters = sum(1 for ch in letters if _SINHALA.search(ch))
    return sinhala_letters >= 1 and sinhala_letters / len(letters) >= 0.5


def has_sinhala_news_text(title: str, excerpt: str | None = None) -> bool:
    if has_sinhala_script(title):
        return True
    combined = f"{title.strip()} {(excerpt or '').strip()}".strip()
    return combined != title.strip() and has_sinhala_script(combined)
